### PR TITLE
deployment watcher: fix goroutine leak when job is purged

### PR DESCRIPTION
**Pull request by tgross**
The deployment watcher on the leader makes blocking queries to detect when the set of active deployments changes. It takes the resulting list of deployments and adds or removes watchers based on whether the deployment is active. But when a job is purged, the deployment will be deleted. This unblocks the query but the query result only shows the remaining deployments.

When the query unblocks, ensure that all active watchers have a corresponding deployment in state. If not, remove the watcher so that the goroutine stops.

Fixes: https://github.com/hashicorp/nomad/issues/19988

### CODE REVIEW

File: nomad/deploymentwatcher/deployments_watcher.go [hunk 1]
```diff
@@ -191,6 +190,21 @@ func (w *Watcher) watchDeployments(ctx context.Context) {
 				w.remove(d)
 			}
 		}
+
+		// Ensure we're not tracking deployments that have been deleted because
```
Comment (Juanadelacuesta) [inline nomad/deploymentwatcher/deployments_watcher.go:194]: If you check for the number of watchers and number of deployments, you could avoid always checking even when there are no changes

Comment (tgross) [reply nomad/deploymentwatcher/deployments_watcher.go:194]: Good catch. That also made me realize that we're not locking the read on `w.watchers` here, so I need to fix that.

Comment (tgross) [reply nomad/deploymentwatcher/deployments_watcher.go]: I don't think we can't optimize this with a length check after all. Imagine the following scenario:

* watchers for deployment A and B exist, len(deployments) == 2, len(watchers) == 2
* A gets purged, B completes (is no longer active), C gets added
* after we call `add`/`remove`, we have len(deployments) == 2, len(watchers) == 2
* so now checking the lengths would be equal and we'd skip removing A

I've pulled this block of code out to a function to make the locking nicer, and I've added a note about why we can't optimize it with the length check.

**Review by shoenig — APPROVED** [review]
LGTM!

**Pull request merged**